cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,735 @@
+                  (order == order::DESCENDING && direction == direction::FOLLOWING),
+       delta     = *row_delta,
+       row_value = begin[i]]() {
+        return subtract ? saturating_sub(row_value, delta) : saturating_add(row_value, delta);
+      }();
+    OrderbyT const offset_value = cuda::std::get<0>(offset_value_overflow);
+    bool const did_overflow     = cuda::std::get<1>(offset_value_overflow);
+    auto const distance         = [preceding    = direction == direction::PRECEDING,
+                           current      = begin + i,
+                           start        = begin + row_info.non_null_start,
+                           end          = begin + row_info.non_null_end,
+                           offset_value = offset_value](auto&& cmp) {
+      if (preceding) {
+        // Search for first slot we can place the offset value
+        return 1 + thrust::distance(thrust::lower_bound(thrust::seq, start, end, offset_value, cmp),
+                                    current);
+      } else {
+        // Search for last slot we can place the offset value
+        return thrust::distance(current,
+                                thrust::upper_bound(thrust::seq, start, end, offset_value, cmp)) -
+               1;
+      }
+    };
+    if (did_overflow) {
+      // If computing the offset row value overflowed then we must
+      // adapt the search comparator. Suppose that the window
+      // direction is PRECEDING and orderby column is ASCENDING. The
+      // offset value is computed as row_value - delta. There are two
+      // cases:
+      // 1. delta is positive: we overflowed towards -infinity;
+      // 2. delta is negative, we overflowed towards +infinity.
+      // For case 1, the saturating minimum value must be
+      // included in the preceding window bound (because -infinity <
+      // min). Conversely, for case 2, the saturating maximum value
+      // must not be included in the preceding window bound (because
+      // max < +infinity). This can be obtained by picking a
+      // bounded_closed (respectively bounded_open) window.
+      // For FOLLOWING windows (with ASCENDING orderby), positive
+      // delta overflows towards +infinity and negative delta towards
+      // -infinity, and the same logic applies: we should include for
+      // positive overflow, but exclude for negative overflow.
+      // Since, when the orderby columns is ASCENDING, delta is
+      // treated with a sign flip, the above also applies in that case.
+      if (*row_delta > DeltaT{0}) {
+        return distance(comparator_t<OrderbyT, bounded_closed>{order});
+      } else {
+        return distance(comparator_t<OrderbyT, bounded_open>{order});
+      }
+    } else {
+      return distance(Comp{order});
+    }
+  }
+};
+
+/**
+ * @brief Functor to dispatch computation of clamped range-based rolling window bounds.
+ *
+ * @tparam WindowType The tag indicating the type of window being computed
+ */
+template <typename WindowType>
```

Comment:
This template parameter could be made runtime by accepting the `range_window_type` variant rather than the type as a parameter here. However, it won't change the compile time because we must instantiate all of the device kernels separately anyway.